src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -21706,6 +21706,484 @@ void M(string? s)
+            CompileAndVerify(comp2);
+        }
+
+        [Fact]
+        public void NotNullIfNotNull_Return_NullableInt()
+        {
+            var lib_cs =
+@"using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    [return: NotNullIfNotNull(""p"")] public int? M(int? p) => throw null!;
+}";
+            var source = @"
+class D
+{
+    static void M(C c, int? p1, int? p2)
+    {
+        _ = p1 ?? throw null!;
+        c.M(p1).Value.ToString();
+        c.M(p2).Value.ToString(); // 1
+    }
+}";
+            var expected = new[]
+            {
+                // (8,9): warning CS8629: Nullable value type may be null.
+                //         c.M(p2).Value.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullableValueTypeMayBeNull, "c.M(p2)").WithLocation(8, 9)
+            };
+
+            var lib = CreateNullableCompilation(new[] { NotNullIfNotNulltAttributeDefinition, lib_cs });
+            var comp = CreateNullableCompilation(source, references: new[] { lib.EmitToImageReference() });
+            comp.VerifyDiagnostics(expected);
+
+            var comp2 = CreateNullableCompilation(new[] { source, lib_cs, NotNullIfNotNulltAttributeDefinition });
+            comp2.VerifyDiagnostics(expected);
+            CompileAndVerify(comp2);
+        }
+
+        [Fact]
+        public void NotNullIfNotNull_Return_MultipleParameters()
+        {
+            var lib_cs =
+@"using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    [return: NotNullIfNotNull(""p""), NotNullIfNotNull(""p2"")] public string? M(string? p, string? p2) => throw null!;
+}";
+            var source = @"
+class D
+{
+    static void M(C c, string? p1, string? p2)
+    {
+        _ = p1 ?? throw null!;
+        c.M(p1, p1).ToString();
+        c.M(p1, p2).ToString();
+        c.M(p2, p1).ToString();
+        c.M(p2, p2).ToString(); // 1
+    }
+}";
+            var expected = new[]
+            {
+                // (10,9): warning CS8602: Dereference of a possibly null reference.
+                //         c.M(p2, p2).ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c.M(p2, p2)").WithLocation(10, 9)
+            };
+
+            var lib = CreateNullableCompilation(new[] { NotNullIfNotNulltAttributeDefinition, lib_cs });
+            var comp = CreateNullableCompilation(source, references: new[] { lib.EmitToImageReference() });
+            comp.VerifyDiagnostics(expected);
+
+            var comp2 = CreateNullableCompilation(new[] { source, lib_cs, NotNullIfNotNulltAttributeDefinition });
+            comp2.VerifyDiagnostics(expected);
+            CompileAndVerify(comp2);
+        }
+
+        [Fact]
+        public void NotNullIfNotNull_Return_DuplicateParameters()
+        {
+            var lib_cs =
+@"using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    [return: NotNullIfNotNull(""p2""), NotNullIfNotNull(""p2"")] public string? M(string? p, string? p2) => throw null!;
```

Comment:
Should this be a warning? Feels like it should be, and if it is we need to do it now.